perception/autoware_tensorrt_bevdet/src/bevpool_plugin.cu

```diff
@@ -0,0 +1,364 @@
+{
+  return DataType::kFLOAT;
+}
+
+DimsExprs BEVPoolPlugin::getOutputDimensions(
+  int32_t outputIndex, const DimsExprs * inputs, int32_t nbInputs,
+  IExprBuilder & exprBuilder) noexcept
+{
+  // input[0] depth            b*n x d x h x w
+  // input[1] feat             b*n x c x h x w
+
+  // input[2] ranks_depth      m
+  // input[3] ranks_feat       m
+  // input[4] ranks_bev        m
+  // input[5] interval_starts  k
+  // input[6] interval_lengths k
+
+  // out[0]   bevfeat          b x c x h x w
+
+  DimsExprs ret;
+  ret.nbDims = inputs[1].nbDims;
+  ret.d[0] = exprBuilder.operation(
+    DimensionOperation::kFLOOR_DIV, *inputs[1].d[0], *exprBuilder.constant(m_.n));
+  ret.d[1] = inputs[1].d[1];
+  ret.d[2] = exprBuilder.constant(m_.bev_h);
+  ret.d[3] = exprBuilder.constant(m_.bev_w);
+
+  return ret;
+}
+
+bool BEVPoolPlugin::supportsFormatCombination(
+  int32_t pos, const PluginTensorDesc * inOut, int32_t nbInputs, int32_t nbOutputs) noexcept
+{
+  // depth
+  if (pos == 0) {
+    return (inOut[pos].type == DataType::kFLOAT || inOut[pos].type == DataType::kHALF) &&
+           inOut[pos].format == TensorFormat::kLINEAR;
+  } else if (pos == 1) {  // feat
+    return inOut[0].type == inOut[1].type && inOut[pos].format == TensorFormat::kLINEAR;
+  } else if (pos == 7) {  // out
+    return (inOut[pos].type == DataType::kFLOAT || inOut[pos].type == DataType::kHALF) &&
+           inOut[pos].format == TensorFormat::kLINEAR;
+  } else {
+    return inOut[pos].type == DataType::kINT32 && inOut[pos].format == TensorFormat::kLINEAR;
+  }
+  return false;
+}
+
+void BEVPoolPlugin::configurePlugin(
+  const DynamicPluginTensorDesc * in, int32_t nbInputs, const DynamicPluginTensorDesc * out,
+  int32_t nbOutputs) noexcept
+{
+  return;
+}
+
+size_t BEVPoolPlugin::getWorkspaceSize(
+  const PluginTensorDesc * inputs, int32_t nbInputs, const PluginTensorDesc * outputs,
+  int32_t nbOutputs) const noexcept
+{
+  return 0;
+}
+
+int32_t BEVPoolPlugin::enqueue(
+  const PluginTensorDesc * inputDesc, const PluginTensorDesc * outputDesc,
+  const void * const * inputs, void * const * outputs, void * workspace,
+  cudaStream_t stream) noexcept
+{
+  // input[0] == depth            b*n x d x h x w
+  // input[1] == feat             b*n x c x h x w
+  // input[2] == ranks_depth      m
+  // input[3] == ranks_feat       m
+  // input[4] == ranks_bev        m
+  // input[5] == interval_starts  k
+  // input[6] == interval_lengths k
+
+  int channel = inputDesc[1].dims.d[1];
+  int n_intervals = inputDesc[5].dims.d[0];
+  int map_size = m_.bev_h * m_.bev_w;
+  int img_size = inputDesc[0].dims.d[2] * inputDesc[0].dims.d[3];
+
+  dim3 grid(GET_BLOCKS(n_intervals * channel));
+  dim3 block(NUM_THREADS);
+
+  switch (int(outputDesc[0].type)) {
+    case int(DataType::kFLOAT):
+      if (inputDesc[0].type == DataType::kFLOAT) {
+        // fp32 fp32
+        bev_pool_v2_kernel<float, float><<<grid, block, 0, stream>>>(
```

Comment:
Within the `enqueue` function, the numerous `reinterpret_cast` calls should be reviewed for potential safety issues. Consider using static asserts to verify types at compile time to improve robustness and prevent runtime errors.  For example, you could add assertions or static_assert to ensure input and output types match expectations.